ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifier.java

```diff
@@ -0,0 +1,289 @@
+  }
+
+  public void onUpdatePreparableProposers(
+      Collection<BeaconPreparableProposer> beaconPreparableProposers) {
+    LOG.debug("onUpdatePreparableProposers {}", beaconPreparableProposers);
+
+    recentChainData
+        .getCurrentSlot()
+        .ifPresent(
+            currentSlot ->
+                updatePreparableProposers(
+                    beaconPreparableProposers,
+                    attestationDueHasBeenCalledForSlot(currentSlot)
+                        ? currentSlot.plus(1)
+                        : currentSlot));
+  }
+
+  private boolean attestationDueHasBeenCalledForSlot(UInt64 slot) {
+    return maybeLastAttestationDueSlot
+        .map(lastAttestationDueSlot -> lastAttestationDueSlot.isGreaterThanOrEqualTo(slot))
+        .orElse(false);
+  }
+
+  public void onForkChoiceUpdated(
+      Bytes32 optimisticHeadBeaconBlockRoot, ForkChoiceState forkChoiceState) {
+    LOG.debug("onForkChoiceUpdated {}, {}", optimisticHeadBeaconBlockRoot, forkChoiceState);
+
+    Optional<Bytes32> currentValidatedHead =
+        recentChainData.getChainHead().map(StateAndBlockSummary::getRoot);
+    boolean isForkChoiceStateOnValidatedHead =
+        currentValidatedHead.isPresent()
+            && optimisticHeadBeaconBlockRoot.equals(currentValidatedHead.get());
+
+    currentForkChoiceUpdatedInfo.forkChoiceState = Optional.of(forkChoiceState);
+    currentForkChoiceUpdatedInfo.isOnValidatedHead = isForkChoiceStateOnValidatedHead;
+
+    recentChainData
+        .getCurrentSlot()
+        .ifPresent(
+            (slot) -> {
+              UInt64 targetSlot = slot.plus(1);
+              callExecutionEngineForkChoiceUpdated(getPayloadAttributes(targetSlot), targetSlot);
+            });
+  }
+
+  public void onAttestationsDue(UInt64 slot) {
+    LOG.debug("onAttestationsDue {}", slot);
+
+    maybeLastAttestationDueSlot = Optional.of(slot);
+
+    UInt64 targetSlot = slot.plus(1);
+
+    if (currentForkChoiceUpdatedInfo.hasNotBeenCalledForSlot(targetSlot)) {
+
+      Optional<PayloadAttributes> payloadAttributes = getPayloadAttributes(targetSlot);
+
+      if (payloadAttributes.isPresent()) {
+        callExecutionEngineForkChoiceUpdated(payloadAttributes, targetSlot);
+      }
+    }
+  }
+
+  public CurrentForkChoiceUpdatedInfo getCurrentForkChoiceUpdatedInfo() {
+    return currentForkChoiceUpdatedInfo;
+  }
+
+  private void updatePreparableProposers(
+      Collection<BeaconPreparableProposer> beaconPreparableProposers, UInt64 currentSlot) {
+
+    final int maxProposerSeenSlot = spec.getSlotsPerEpoch(currentSlot) * MAX_PROPOSER_SEEN_EPOCHS;
+    boolean newProposerIndexAdded = false;
+
+    // remove expired proposers
+    proposerIndexLastSeenSlotAndFeeRecipient =
+        proposerIndexLastSeenSlotAndFeeRecipient.entrySet().stream()
+            .filter(p -> p.getValue().lastSeenSlot.compareTo(currentSlot) < maxProposerSeenSlot)
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```

Comment:
We shouldn't be recreating the entire map each time here.  I think we can just use:
```suggestion
    // remove expired proposers
    proposerIndexLastSeenSlotAndFeeRecipient
        .values()
        .removeIf(lastSeen -> lastSeen.lastSeenSlot.isGreaterThanOrEqualTo(maxProposerSeenSlot));
```

Also interestingly the PR to the REST API requires the beacon node to persist these values until restart - I've added a comment suggesting we allow the node to drop them after a period, so for now I'd stick with expiring after 2 epochs as the code is doing.